RFC/src/RFC-0160_BlockSerialization.md

```diff
@@ -0,0 +1,287 @@
+THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+
+## Language
+
+The keywords "MUST", "MUST NOT", "REQUIRED", "SHALL", "SHALL NOT", "SHOULD", "SHOULD NOT", "RECOMMENDED",
+"NOT RECOMMENDED", "MAY" and "OPTIONAL" in this document are to be interpreted as described in
+[BCP 14](https://tools.ietf.org/html/bcp14) (covering RFC2119 and RFC8174) when, and only when, they appear in all capitals, as
+shown here.
+
+## Disclaimer
+
+This document and its content are intended for information purposes only and may be subject to change or update
+without notice.
+
+This document may include preliminary concepts that may or may not be in the process of being developed by the Tari
+community. The release of this document is intended solely for review and discussion by the community of the
+technological merits of the potential system outlined herein.
+
+## Goals
+
+The aim of this Request for Comment (RFC) is to specify the binary serialization of:
+
+1. a mined Tari block
+1. a Tari block mining template
+
+This is to facilitate interoperability of mining software and hardware.
+
+## Related Requests for Comment
+
+- [RFC-0131: Full-node Mining on Tari Base Layer](RFC-0131_Mining.md)
+
+## Specification
+
+By reviewing the [block and mining template fields](#tari-block-and-mining-template-data-types) below, we have the following underlying data types for serialization:
+
+1. `bool`
+1. `u8`
+1. `u16`
+1. `u32`
+1. `Vec<u8>`
+
+For 1. to 4. and all numbers, [Base 128 Varint] encoding MUST be used.
+
+From the Protocol Buffers documentation:
+
+> Varints are a method of serializing integers using one or more bytes. Smaller numbers take a smaller number of bytes. Each byte in a varint, except the last byte, has the most significant bit (msb) set – this indicates that there are further bytes to come. The lower 7 bits of each byte are used to store the two's complement representation of the number in groups of 7 bits, least significant group first.
+
+For 5., the dynamically sized `Vec` array type, the encoded array MUST be preceded by a number indicating the length of the array. This length MUST also be encoded as a varint. By prepending the length of the array, the decoder knows how many elements to decode as part of the sequence.
+
+[base 128 varint]: https://developers.google.com/protocol-buffers/docs/encoding#varints
+
+## Block field ordering
+
+Using this varint encoding, all fields of the complete block MUST be encoded in the following order:
+
+1. Version
+2. Height
+3. Previous block hash
+4. Timestamp
+5. Output Merkle root
+6. Witness Merkle root
+7. Output Merkle mountain range size
+8. Kernel Merkle root
+9. Kernel Merkle mountain range size
+10. Input Merkle root
+11. Total kernel offset
+12. Total script offset
+13. Nonce
+14. Proof of work algorithm
+15. Proof of work supplemental data
+16. Transaction inputs - for each input:
```

Comment:
Perhaps we should discuss the input, output and kernel ordering here as well. The list of them are ordered and not just in some random order.

Reply:
Good point I missed that, thanks @SWvheerden